measureme/src/serialization.rs

```diff
@@ -11,70 +14,172 @@ impl Addr {
     }
 }
 
-pub trait SerializationSink: Sized + Send + Sync + 'static {
-    fn from_path(path: &Path) -> Result<Self, Box<dyn Error + Send + Sync>>;
-
-    /// Atomically write `num_bytes` to the sink. The implementation must ensure
-    /// that concurrent invocations of `write_atomic` do not conflict with each
-    /// other.
-    ///
-    /// The `write` argument is a function that must fill the output buffer
-    /// passed to it. The output buffer is guaranteed to be exactly `num_bytes`
-    /// large.
-    fn write_atomic<W>(&self, num_bytes: usize, write: W) -> Addr
-    where
-        W: FnOnce(&mut [u8]);
+#[derive(Debug)]
+pub struct SerializationSink {
+    data: Mutex<Inner>,
+}
+
+/// The `BackingStorage` is what the data gets written to.
+trait BackingStorage: Write + Send + Debug {
+    fn drain_bytes(&mut self) -> Vec<u8>;
```

Comment:
Should we #[cfg(test)] this or would that be too weird?

Reply:
I just tried it but it's complicated because `analyzeme` uses these things in its tests too -- but Cargo does not compile the dependencies of crate X with `--test` or `--cfg test` when it compiles crate X with `--test`. I suspect one can somehow use Cargo features to get this to work (although I couldn't find an example of this in the Cargo book).

Reply:
Ah ok. Let's add a doc comment to this saying it shouldn't be used except by tests because it will panic.

Reply:
Done.